Test-ception

knyazer opened the pull request:
Good evening from a Finnish forest,


This fixes #15 

I ignored everything you all guys mentioned in there, because I didn't feel like dealing with generators. Instead, the idea is to add a custom ```fixture_setup``` that replaces fixtures failing typechecking with a little class, and then detects these little classes in ```pyfunc_call```.

I thought initially this won't work with recursive fixtures, but all seems good and ```pyfunc_call``` resolves all the fixture calls in the correct order which is very nice. This might hit some corner cases, but I don't have enough braincells to think about any.

What else... Ah, yes, I removed the dependency on the coverage - because it was breaking pytest? I have no idea why, but who needs coverage or "code quality" anyway. Also added uv lockfile so that people using uv will have easier time developing.

Added some explanations into README. Btw, the generator cornercase now should not be hit ever, since for it to be triggered you need to make a fixture that is a generator _by itself_ which I don't think is a valid pytest syntax? Haven't checked though.

@leycec please take a look

knyazer commented:
Err, 3.13 is having fun... I'll take a look now

tusharsadhwani commented:
Sounds like a very neat idea from what I've seen so far. I'll be able to review this Tuesday, other comments are very welcome :^)

Can you drop `uv.lock` and `.python-version` from the changes?

leycec commented:
@tusharsadhwani: You still live! I hope you had an amazing summer – and that your super-busy schedule has quieted down to something resembling normalcy. It's so fun to continue hearing from you. :heart_eyes: 

It sure was a sweltering summer up here in Canada. For one frightening moment in time, all of Canada transformed into a gigantic hot-house jungle straight out of 2000AD. It was *insane*. Rabid jungle insects the size of your fist. Rabid jungle trees growing whole feet in a day. Torrential monsoon rain the same time every day you could set your command-line clock by. Where did the real Canada go!? :hot_face: 

Oh, and...

> Can you drop `uv.lock` and `.python-version` from the changes?

**I sadly second this.** These are otherwise phenomenal changes, @knyazer. But devtool-specific cache files like this are *guaranteed* to both:

* Break at some future date and...
* Ignite a conflagration of hatred from [Dependabot](https://docs.github.com/en/code-security/getting-started/dependabot-quickstart-guide#about-dependabot), which I'm *just* about to enable for both this repo and the [adjacent `ipython-beartype` repo](https://github.com/beartype/ipython-beartype).

@knyazer: Given that, would you mind adding both `uv.lock` and `.python-version` to this repo's top-level `.gitignore`? You can then run with those files on your local devbox *without* introducing spurious Dependabot conflicts. Does that sound reasonable? If not, feel free to pump your fist angrily in the direction of Canada.

## And Now, The Moment We've All Been Waiting For...

Let's code review this phenomenal piece of the `pytest-beartype` puzzle! :jigsaw:

leycec commented on .python-version:
As @tusharsadhwani suggests, let's instead **(A)** remove this file from this changeset and **(B)** add `.python-version` to this repo's `.gitignore` file. `.python-version` is kinda weird and basically guaranteed to become desynchronized with modern Python versions at some later date, which could be as soon as tomorrow. The future: *it's coming faster than you want.* :smiling_face_with_tear:

leycec commented on README.md:
**Phenomenal.** You went above and beyond the call of documentation duty here, @knyazer. Thanks so much for exhaustively describing this new functionality. I am shock face. :open_mouth:

leycec commented on conftest.py:
```diff
@@ -0,0 +1,5 @@
+import sys
+
+sys.dont_write_bytecode = True  # bytecode breaks stuff cuz what i'm doing is magic
```
**This is so weird.** I accept that we *might* need to do this. But do we *really* need to do this? If so, let's comment this weirdness in a bit more detail. Specifically:

* What exactly is the magic that we are doing that requires avoiding bytecode-compilation?
* What exactly happens if we *don't* do this?
* Does `pytest` start throwing up some disgusting errors that make us want to give up coding for the year... again?

Ideally, we really *should* be compiling bytecode everywhere. Thankfully, this only impacts the test suite. So, it's not a huge deal. We can do this if we really need to do this. Still, it's kinda weird and suspicious. I smell a bad code smell. :pig_nose:

knyazer replied:
Well, I don't feel like it's bad code: it's more of the "fix pytest" kinda code. Indeed, if I remove the line pytest goes a bit crazy; I think because "pytester" tries to use the old bytecode, and then blows up. I'll write a more exhaustive comment, since you asked (originally I didn't feel like doing so, since it seems like something that might be python/pytest version dependent)

leycec replied:
**Gotcha.** Indeed, this sounds like bonafide `pytester` bugs. That's out of our wheelhouse, which is the best kind of bugs. If you wouldn't mind commenting this a bit, that would be amazing!

Thanks so much for the extreme efforts you've gone here, too. I know my code review is pedantic and thus annoying... I only ask that you please *bear* with me. :bear:  :pray:

leycec commented on src/pytest_beartype/__init__.py:
@knyazer: **SHEER GENIUS.** I stand in awe before a keyboard that anyone could possibly code this. Seriously! Who slings `pytest` plugin API code like this? The `pytest` plugin API is barely even documented. The documentation is literally the `pytest` codebase at this point. It's almost worse than @beartype documentation, which no one thought was possible. :sob: 

I *never* would have thought to separate the fixture type-checking phase from the test type-checking phase, conjoined by this wild `_BeartypeFixtureFailure` interchange format that you then use to intercommunicate fixture failures from the fixture type-checking phase to the test type-checking phase. This weekend, you have ascended to `pytest` Godhood.

![](https://media1.tenor.com/m/MnnB5y5mvaQAAAAd/worried.gif)
<sup>random middle-aged salary-man goes wide-eyed with a sudden realization: "I no longer understand what is going on."</sup>

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -3,6 +3,8 @@
 from __future__ import annotations
```
**Uh oh.** `pytest-beartype` *definitely* shouldn't be enabling PEP 563 via `from __future__ import annotations`. PEP 649 and 749 have officially deprecated PEP 563. In six years, `from __future__ import annotations` will begin raising fatal `SyntaxError` exceptions. For further details that will make your eyes sting with pain, see also:

* The [@beartype `0.22.0rc0` release notes](https://github.com/beartype/beartype/discussions/551).

Would you mind removing this self-destructive pragma, @knyazer? I beg this of you! :pray:

knyazer replied:
Sure :)

leycec replied:
**Wondrous!** @beartype continues saluting you. :vulcan_salute:

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+    """
+    Apply the beartype decorator to all collected test functions if --beartype-check-tests is enabled.
+    """
+    beartype_tests_enabled = config.getini("beartype_check_tests") or config.getoption(
```
**Ho, ho!** [Don't Repeat Yourself (DRY)](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) violation here. This same logic is repeated in the `pytest_pyfunc_call()` function below. Humour Grandpa @leycec, but would you mind refactoring this out into a private tester function? Perhaps something resembling:

```python
def _is_pytest_config_beartype_check_tests(config: "pytest.Config") -> bool:
    return (
        config.getini("beartype_check_tests") or
        config.getoption("beartype_check_tests", False)
    )
```

Also, a related important question for both you and @tusharsadhwani:

* **Should `--beartype_check_tests` default to `True` instead of `False`?**

This is globally useful type-checking. Ideally, *all* `pytest-beartype` users will want to enable this option. It seems reasonable that `pytest` tests and fixtures should *always* be type-checked by default. Does that make sense? If so, we should consider revising the above private tester implementation to resemble:

```python
def _is_pytest_config_beartype_check_tests(config: "pytest.Config") -> bool:
    return (
        config.getini("beartype_check_tests") or
        config.getoption("beartype_check_tests", True)  # <-- the magic begins today
    )
```

knyazer replied:
Yup, that is a **very** good idea. Thank you, bear god.

My default instinct is to keep backwards compatibility, but I feel like considering the number of users of pytest_beartype we might not need to care. I'll enable the option by default (and yeah, maybe it is indeed an "expected" behaviour, so we should enable it anyway)

leycec replied:
> My default instinct is to keep backwards compatibility

To be sure, this could temporarily break somebody's downstream test suite. Usually, that's bad. In this case, that's *mostly* fine. Either:

* **(A)** That test suite was broken anyway and they didn't know it (which is awful), but now they do (which is good).
* **(B)** That test suite wasn't actually broken. It just had bad type hints (which isn't great), but at least they know about it now (which is good).

Basically, users should be informed that something is bad in their test suite. Right? That's kinda the whole point of this plugin. Thankfully, users who don't care there is something bad in their test suite can just disable this option.

We're not the bad guys here! Oh, Gods... I think we might be the bad guys here. :joy:

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+    for item in items:
+        if isinstance(item, pytest.Function):
+            # Wrap the test function with beartype
+            item.obj = beartype(item.obj)
```
@adamtheturtle: Your genius combined with @knyazer's genius to create a terrifying new genius. :exploding_head:

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+    instead of "error" (which usually indicates an internal pytest error, which is wrong
+    in this case).
+    """
+    beartype_tests_enabled = request.config.getini(
```
**Ho, ho!** DRY violation intensifies. See comment above. :stuck_out_tongue:

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+    # TODO: force tiny cub @knyazer or Bear God @leycec to fix this when something is done with it
+    # This also should not really happen, like ever, since I don't understand what it means
+    # for a fixture _function_ to be a generator, and I don't think pytest supports it
+    if inspect.isgeneratorfunction(fixturedef.func):
```
Publicizing my ancient shame for everybody to see, huh? I love it. Thanks for checking this pernicious edge case. If this doesn't prod me into doing something about beartype/beartype#423, then nothing will. So... I guess nothing will. :rofl:

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+        try:
+            # Apply beartype decoration here, at call time
+            beartype_decorated = beartype(original_func)
+            return beartype_decorated(*args, **kwargs)
```
**Ho, ho!** Potential performance regression here. The `@beartype` decorator is *incredibly* slow. The `beartype_decorated()` function call should be much faster in comparison. This doesn't matter for session-scoped fixtures, of course; they're only called once, so it doesn't particularly matter that you're re-calling the `@beartype` decorator once per fixture call.

This totally does matter for test-scoped fixtures, though; they're called once per test, obviously. Re-calling the `@beartype` decorator once per fixture call suddenly matters there – especially for large test suites with crazy custom test-scoped fixtures firing off all over the place.

In theory, we should be able to optimize away these concerns by separating out the application of the `@beartype` decorator from the calling of the decorated `beartype_decorated()` fixture ala:

```python
    try:
        # Apply beartype decoration here at fixture definition time.
        beartype_decorated = beartype(original_func)

        # Create a wrapper that catches beartype errors
        @functools.wraps(original_func)
        def beartype_fixture_wrapper(*args: Any, **kwargs: Any) -> Any:
            try:
                return beartype_decorated(*args, **kwargs)
            except BeartypeException as e:
                # Return sentinel object instead of raising
                return _BeartypeFixtureFailure(fixturedef.argname, e)
    except BeartypeException as e:
        # Create a wrapper that returns a sentinel object encapsulating this
        # early @beartype decorator-time exception.
        @functools.wraps(original_func)
        def beartype_fixture_wrapper(*args: Any, **kwargs: Any) -> Any:
            # Return sentinel object instead of raising
            return _BeartypeFixtureFailure(fixturedef.argname, e)
```

Admittedly, that's messy. I acknowledge the ugliness inherent in fast code. More importantly, though, that is *fast.* Fast >>>>>>>> messy.

Feel free to clean that up any if you can. For example, you might be able to define the `def beartype_fixture_wrapper()` closure just once with a bit of cleverness. I shrug! I shrug so hard! :shrug:

knyazer replied:
Oh oops, I forgot to resolve this one :) Sorry

Will fix it up in a few days

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+        self.error = error
+
+
+def pytest_fixture_setup(
```
**GOD-TIER.** This is some solid `pytest` tomfoolery right here. I only have applause to give. Please come to Canada someday. Let's recreate this epic scene of debauched maple syrup addiction.

![](https://media1.tenor.com/m/Y1OW5j-8_xoAAAAd/super-troopers-syrup.gif)
<sup>left: *@leycec*. right: *@knyazer*. middle-left: *@tusharsadhwani, dumbfounded*.</sup>

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+    This hook runs during the actual test execution, allowing us to fail
+    the test (not the setup) when beartype fixture violations are detected.
+    """
+    beartype_tests_enabled = pyfuncitem.config.getini(
```
**Ho, ho!** DRY violation intensifies. See comment above.

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+
+    # Check all fixture values for beartype failures before calling the test
+    for argname in pyfuncitem.fixturenames:
+        fixture_value = pyfuncitem._request.getfixturevalue(argname)
```
**Sheer madness.** How did you figure this special darkness out, @knyazer!? None of these `pytest` internals are even publicly documented. :weary:

knyazer replied:
I once was writing a mutation testing plugin for pytest, and this led me to learn some of its internals. Those were the days... I remember writing so much ugly code monkey patching pytest internals, only for it to break in unrelated places :smiling_face_with_three_hearts: 

I hope to never do that again..

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +142,129 @@ class BeartypePytestWarning(BeartypeWarning):
+    for argname in pyfuncitem.fixturenames:
+        fixture_value = pyfuncitem._request.getfixturevalue(argname)
+        if isinstance(fixture_value, _BeartypeFixtureFailure):
+            pytest.fail(
```
**Ho, ho!** Awesome, but we're totally dropping the associated exception traceback on the floor. Ideally, we should also be printing that somehow. The [poorly documented `pytest.fail()` API](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-fail) appears to suggest that the exception traceback can be printed by additionally passing the **deprecated** `msg` parameter as follows:

```python
            from traceback import format_tb
            pytest.fail(
               f"Fixture '{fixture_value.fixture_name}' failed beartype validation: {fixture_value.error}",
                pytrace=False,
                msg=format_tb(fixture_value.error.__traceback__),  # <-- eldritch darkness
            )
```

Does that actually work, though? And isn't `msg` deprecated, anyway? No idea, bro. `pytest` madness is a special darkness I care not to understand. :vomiting_face:

knyazer replied:
I think the new-and-modern way of doing it is to add traceback at the new line of the content string. I'm not sure. I'll take a look.

leycec replied:
**Super-duper.** Anything you come up with is perfect! Thanks for digging deep into the wacky `pytest` API. It's an undocumented jungle out there.

leycec replied:
**Fantastic!** Looks great to me. Thanks for resolving this so fast. My eyes are watering with joy. 🥹

leycec commented on tests/pytest_beartype_test.py:
```diff
@@ -6,31 +6,6 @@
 import pytest_beartype
 
 
-def test_pytest_addoption() -> None:
```
Everyone hates this test now, huh? Sure! No worries. We can definitely remove this unit test if you like.

Is it doing any harm, though? Is it eating babies while guzzling maple syrup in the autumnal darkness? Is this unit test truly the embodiment of all that is corrupt and defiled in this world? Can you explain why this unit test is suddenly The Dark One Arisen? :neckbeard:

knyazer replied:
Yeah, this is one of the tests that is "doing nothing and being annoying", or at least I feel so. It checks that pytest adds/has the correct options, but it does not check that pytest actually does anything about these options, thus luring you into a false sense of security. And as soon as you modify the logic of the main implementation a tiny bit (e.g. by adding a new option, or changing a type of an old option), this test breaks.

And like, all the relevant code that this test tests is about 4 lines of code in the plugin definition. Like, this feels very much like one of these "yeah we need 100% coverage, otherwise our test suite is bad" type of things.

I actually feel less strongly about it right now, but I still think this test is useless.

leycec replied:
> Like, this feels very much like one of these "yeah we need 100% coverage, otherwise our test suite is bad" type of things.

I feel attacked. :joy_cat: 

You're spot on, though. You were right to remove filler like this. A test that increases coverage without meaningfully testing anything just lulls everyone into a false sense of: "This plugin rocks!"

Which it *does*, of course. This plugin does rock, entirely thanks to you and @tusharsadhwani. But not because of fluff like this. Let's mark this nonsense as happily resolved.

leycec commented on tests/pytest_beartype_test.py:
```diff
@@ -81,3 +56,29 @@ def test_bad_weather_with_beartype_violations() -> None:
+
+def test_pytest(pytester):
+    """Test that specific test functions fail when beartype-check-tests option is used."""
+    # TODO: automatically discover al the files under pytest_tests and copy them
```
**Brilliant!** As an addendum that you *definitely* don't need to address now (or ever, really), we might consider eventually trying to discover all the files under `pytest_tests` in a new custom fixture that we define: e.g.,

```python
from pathlib import Path

@pytest.fixture(scope='session'):
def beartype_pytest_tests() -> list[Path]:
    yield Path('tests/pytest_tests').rglob('*.py')

def test_pytest(pytester, beartype_pytest_tests):
    ...
```

Should yield joy, maybe? No idea. Thankfully, this isn't that critical. Feel free to ignore this and just read blood-curdling manga all week instead. `\o/`

knyazer replied:
Oh, that's very smart! I somehow just gave up when trying to come up with a way to discover the files under pytest_tests automatically, because pytester does something very very very ugly under the hood, and I didn't want to deal with it. But your solution should work. I suppose.

leycec replied:
I'm detecting a common thread through this code review: **`pytester` is ugly.** At least it's not me for once. :laughing: 

If the above workaround doesn't work out, that's totally fine. At least we tried to do the right thing. And that's all anyone can ask in 2025. Thanks yet again for looking into all this QA madness.

leycec commented on tests/pytest_tests/test_function_decoration.py:
```diff
@@ -0,0 +1,73 @@
+
+@pytest.mark.skip
+def test_function_with_type_violation() -> None:
+    # This does not cause beartype violation, because, ehhh, that is like impossible to check
```
**Ha, ha!** Through the terrifying power of `beartype.claw` import hooks, we totally *could* begin type-checking embedded classes and callables at some later date. I'd actually be a huge fan of that. :sweat_smile: -> :sweat: 

If we do ever try something like that, we'll have to be careful to avoid stepping on the toes of the `pytest_fixture_setup()` function. Actually, maybe it'll all just work out if we blindly point `beartype.claw.beartype_package()` at user test suites and hope for the best? I. Know. Nothing.

knyazer replied:
Nope, I checked, it does not work - beartype just happily ignores the existance of embedded functions; this is one of the tests that I wrote and it was failing, so I marked it with "skip" for future reference :)

leycec replied:
**Woah.** You checked into everything. You have an answer for everything. You're a `pytest` guru who walks among us. My uneducated assessment is that `beartype.claw` import hooks conflict with `pytest` import hooks as follows (in order):

1. First, `beartype.claw` import hooks are correctly producing type-checked bytecode that type-checks embedded functions.
2. Next, `pytest` import hooks then silently overwrite that bytecode with their own `pytest`-specific bytecode – discarding all of @beartype's hard work.

Python *really* needs a new PEP standard for import hooks. Currently, there's no standard. It's complete chaos. Competing import hooks either ignore each other or beat each other up.

**Oh, well.** There's not much we can do here. 2025: *why you gotta be like this?* :face_exhaling:

leycec commented on uv.lock:
As @tusharsadhwani suggests, let's instead **(A)** remove this file from this changeset and **(B)** add `uv.lock` to this repo's `.gitignore` file. It's probably best to avoid packager-specific lock files like this if we can help it. :face_in_clouds:

leycec commented:
**tl;dr:** God-tier coding right here. I'm a huge fan of your work, @knyazer. You are the Big Boss of `pytest`. :hugs:

knyazer commented:
I am very sorry about including uv files. It is totally my fault, and I _bear_ all of the responsibility. I will demote myself from the "cub" status to the "infant bear" status. 

This won't happen again! I promise.. <sup>unless I forget about it :)</sup>

On a serious note, thank you for the kind words Mr Leycec!

leycec commented:
**You're amazing!** Especially because you're apparently only an infant bear now. If this is what your infant powers can already do, there truly are no limits for you. Even baby @knyazer is the Terror of QA. :baby: -> :bear: 

![](https://media1.tenor.com/m/aZHhblVXq7gAAAAd/polar-bear-baby.gif)
<sup>*baby polar @knyazer bear gratefully accepts a massage from the @beartype userbase as this latest PR exhausts his infant powers*</sup>

knyazer commented:
@tusharsadhwani please take a look when you are available

leycec commented:
@knyazer: @tusharsadhwani is sadly indisposed due to work and having an awesome life. So proud of his accomplishments.

Thankfully, @leycec is on the job for once! That's me. I don't know why I'm talking about myself in the third person. Yet, here we are. I *just* excreted @beartype ~~`0.22.0`~~ ~~`0.22.1`~~ `0.22.2` and now find myself with a shocking amount of something called "free time." It's not a thing I am familiar with. Yet, it already feels wonderful. Therefore, allow me to now acknowledge...

## I AM A SUPREME IDIOT.

I reviewed your amazing changeset **over two weeks ago** with a honking truckload of comments. I wondered why you never replied, but was too busy squabbling with Python 3.14 to devote time. In short, I was dumb. I didn't realize that the *insanely overengineered* GitHub UI for code reviews requires explicitly pushing the hidden **Submit changes** button in the right-hand corner of the "Files changed" tab. Until that *insanely overengineered* button is pushed, a code review is labelled `Pending`. `Pending` code reviews are silently hidden from everyone except the code reviewer, who naturally assumes that everyone is grumpy and just ignoring them.

**What a trash UI.** There's even a [whole GitHub Community discussion on how trash this UI is](https://github.com/orgs/community/discussions/10369). Apparently, many GitHub code reviewers have been waiting literally years for their code review to be acknowledged – only to belatedly realize that it's been `Pending` and thus silently hidden. For years. Lots of user anger in that thread. I channel that anger as well.

Not a fan, GitHub. Not a fan at all. :face_exhaling: 

We can't fix GitHub's broken code review UI, but we can complain about it. This is that complaint. All of which is to say...

## I Humbly Apologize for the Extremely Lengthy Delay

I'm so sorry you didn't receive my code review immediately. That's entirely on me – and GitHub's trash UI. Let us never forget that I tried to do the right thing but failed. Thus, this is me apologizing publicly towards you. You may have assumed that @tusharsadhwani and I were negligent, rude, and/or bad people. In truth:

* @tusharsadhwani is simply a busy-but-great person.
* I am simply a dumb-but-okay person.

The wages of GitHub sin is public shame like this.

![](https://media1.tenor.com/m/a_DFSqheS8kAAAAd/cute-cat.gif)
<sup>*@beartype begs for forbearance, forgiveness, and other words that start with "for"*</sup>

knyazer commented:
Alright, so in conclusion: the bytecode flag was needed for it to work on my slightly unhinged workstation setup (where /tmp and /home are owned by different PCs with different architectures, which made things a bit harder for pytester).

Otherwise, fixed some things, changed the flag to --beartype-ignore-tests and that's about it.

I also dropped Python3.9 from tox, since the support just ended :)

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +140,135 @@ class BeartypePytestWarning(BeartypeWarning):
+    # Tests are checked by default unless --beartype-ignore-tests is passed
+    return not (
+        config.getini("beartype_ignore_tests")
+        or config.getoption("beartype_ignore_tests", False)
```
Thanks so much for preserving DRY with this utility function. Would you mind flipping this `False` to `True` to enable this option by default? Or... is this option already enabled by default somewhere else? My `pytest` skills are shamefully weak. 🥲

leycec replied:
**Nevermind!** I am dumbness incarnate. You already implicitly flipped the boolean by just flipping the *name* and *interpretation* of this option. Clever. Your intellect continues to stun.

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +140,135 @@ class BeartypePytestWarning(BeartypeWarning):
+    if inspect.isgeneratorfunction(fixturedef.func):
+        # The monkeypatch check is wonderful: monkeypatch is an internal pytest fixture that is,
+        # in fact, a generator. Wow. We don't want to emit warnings all the time, so we just ignore it.
+        if fixturedef.argname != "monkeypatch":
```
**Nice catch!** `monkeypatch` sure is one weird fixture. 😓

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +140,135 @@ class BeartypePytestWarning(BeartypeWarning):
+    import functools
+
+    # Only decorate fixtures that haven't been decorated yet
+    if hasattr(fixturedef, "_beartype_decorated"):
```
This is *really* trivial and ignorable, but you could compact these two `if` conditionals above into one super `if` conditional: e.g.,

```python
    # Only decorate fixtures that haven't been decorated yet
    if (
        hasattr(fixturedef, "_beartype_decorated") or
        not _is_pytest_config_beartype_check_tests(request.config)
    ):
        return
```

Anyways. No biggie. Just my autism acting up for no good reason. 😅

leycec commented on src/pytest_beartype/__init__.py:
```diff
@@ -134,3 +140,135 @@ class BeartypePytestWarning(BeartypeWarning):
+    # Store original function
+    original_func = fixturedef.func
+
+    beartype_decorated = beartype(original_func)
```
**...heh.** I wish this was safe. Unfortunately, the `@beartype` decorator frequently raises exceptions in common use cases. That could happen here if a user annotates a fixture with a type hint that is technically valid but *not* yet supported by @beartype. Although @beartype does support almost all type hints, CPython devs are constantly writing new PEP standards for type hints. It's impossible to keep up with those guys. 😫 

In other words, we really do want to wrap this with a `try: ... except: ...` handler of some sort. That's why my initial commentary on this logic looked like *a lot* more gruesome – like this:

```python
    try:
        # Apply beartype decoration here at fixture definition time.
        beartype_decorated = beartype(original_func)

        # Create a wrapper that catches beartype errors
        @functools.wraps(original_func)
        def beartype_fixture_wrapper(*args: Any, **kwargs: Any) -> Any:
            try:
                return beartype_decorated(*args, **kwargs)
            except BeartypeException as e:
                # Return sentinel object instead of raising
                return _BeartypeFixtureFailure(fixturedef.argname, e)
    except BeartypeException as e:
        # Create a wrapper that returns a sentinel object encapsulating this
        # early @beartype decorator-time exception.
        @functools.wraps(original_func)
        def beartype_fixture_wrapper(*args: Any, **kwargs: Any) -> Any:
            # Return sentinel object instead of raising
            return _BeartypeFixtureFailure(fixturedef.argname, e)
```

I know that's some nasty code, but... I can't think of a better way. Sometimes, you just gotta get nasty.

knyazer replied:
sure, if you say so

I have a lot more belief into beartype being awesome, but you are the one to make the decision on this one :)

leycec commented:
*...so ...close.* We are almost there. I can almost taste Shangri-La. Just two minor quibbles left! The finish line is within reach. :finnadie:

leycec commented:
> I have a lot more belief into beartype being awesome, but you are the one to make the decision on this one :)

**Aww!** You're so nice. I too once believed in @beartype. Then 96 issues were opened on the issue tracker. :rofl: 

Thanks so much for accommodating my pedantry with all this. I'm so impressed by your extreme volunteer ethic, your deep understanding of `pytest` internals, and your congenial good nature. Congratulations, @knyazer. You worked *really* hard on this. You made the world a safer place for all of us. Truly, Finland is a nation of the Gods.

For my part, here's what I'm going to do now (in order):

1. **Create a new GitHub-specific Environment.** PyPI Trusted Publishers requires this. Basically, boring (which means important) security boilerplate.
1. **Create a new GitHub Actions-based `.github/workflows/python_release.yml` workflow.** Again, PyPI Trusted Publishers requires this. I'll probably copy over @beartype's own `python_release.yml` workflow, which is proven to (mostly) work, if we ignore the ugly chaos surrounding @beartype's ill-fated `0.22.0` and `0.22.1` releases. :smiling_face_with_tear: 
1. **Publish your *incredible* work as a new `pytest-beartype 0.3.0` version.**
1. **Announce this new version with a new "Announcement" on the [@beartype discussion forums](https://github.com/beartype/beartype/discussions).** It's likely to get more traction over there. More people are likely to see and read it. The announcement will give everybody a heads-up that:
   * `pytest-beartype` just got a *whole* lot more useful.
   * Backwards compatibility could be an issue for existing users with broken type hints in their `pytest` test suites.

Let's merge this immediately, Bear Family! :bear: :bearded_person: :man_beard: :polar_bear:

adamtheturtle commented:
I'd love to use this. Is it blocked on something?

leycec commented:
**So sorry!** I'd love so much for all of our wondrous users [read: you! it's all you!] to immediately start using and banging on this QA hotness, too. Sadly, your supposition is correct. Blockage exists. And not just trivial blockage either. Blockage of the `pytest-asyncio` variety.

Therefore, you're all [cordially invited to the first `pytest-beartype` State of the QA Address](https://github.com/beartype/pytest-beartype/issues/22)! Don't worry. It's only endless walls of incoherent plaintext – like usual from @leycec. :scream:
